### PR TITLE
Support for np.corrcoef

### PR DESCRIPTION
**Pull request by rjenc29**
Support for `np.corrcoef` (first 3 args - the last 2 are deprecated).

Initial commit for CI.

### CODE REVIEW

**Comment by rjenc29** [comment]
I believe the solitary test failure is indicative of a change in numpy behaviour when faced with an extreme value problem:

**Numpy 1.10 (the build which fails)**

```python
import numpy
print(numpy.__version__)

>>> 1.10.4

x = ((1e-100, 1e100), (1e100, 1e-100))
print(numpy.corrcoef(x))

>>> [[ 0. -0.]
     [-0.  0.]]
```

**Numpy 1.11 and later**

```python
import numpy
print(numpy.__version__)

>>> 1.11.3

x = ((1e-100, 1e100), (1e100, 1e-100))
print(numpy.corrcoef(x))

>>> [[ 1. -1.]
     [-1.  1.]]
```

All other tests pass / behaviour looks consistent across versions with regards to the tests which are executed.  

I could of course bump the version gate but thought I would flag it here for info / discussion.

Let me know what you think.

**Comment by rjenc29** [comment]
I've removed the WIP as it's probably about ready for review (a relatively small change, for once!)

**Comment by stuartarchibald** [comment]
Thanks for adding the feature, much appreciated.

RE: https://github.com/numba/numba/pull/3509#issuecomment-439699392, perhaps just mask off the extreme value test for NumPy < 1.11 so that it doesn't run. I'm of the view that it's not worth trying to replicate something from 3 years ago that has since been made considerably more robust. How does that sound?

**Comment by rjenc29** [comment]
Yep, leave it with me - will make it so!

**Comment by rjenc29** [comment]
Bah, looks like one of the jobs timed out

File: docs/source/reference/numpysupported.rst [hunk 1]
```diff
@@ -264,6 +264,7 @@ The following top-level functions are supported:
 * :func:`numpy.concatenate`
 * :func:`numpy.convolve` (only the 2 first arguments)
 * :func:`numpy.copy` (only the first argument)
+* :func:`numpy.corrcoef` (only the 3 first arguments, requires NumPy >= 1.10 and SciPy >= 0.16)
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst]: Perhaps this should have a note like `implementation behaves as per NumPy 1.11+` or similar ?

File: numba/targets/arraymath.py [hunk 2]
```diff
@@ -1427,6 +1427,40 @@ def _prepare_cov_input(m, y, rowvar, dtype, ddof, _DDOF_HANDLER, _M_DIM_HANDLER)
     _DDOF_HANDLER(ddof)
     return _prepare_cov_input_inner(m, y, rowvar, dtype)
 
+def scalar_result_expected(mandatory_input, optional_input):
+    if isinstance(mandatory_input, types.Array) and mandatory_input.ndim == 1:
+        if optional_input in (None, types.none):
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: Can this just be hoisted out all these functions as a bool and reused?:
```opt_is_none = optional_input in (None, types.none)```

File: numba/targets/arraymath.py [hunk 3]
```diff
@@ -1427,6 +1427,40 @@ def _prepare_cov_input(m, y, rowvar, dtype, ddof, _DDOF_HANDLER, _M_DIM_HANDLER)
...
+
+@register_jitable
+def _clip_corr(x):
+    return np.where(np.fabs(x) > 1, np.sign(x), x)
+
+@register_jitable
+def _clip_complex(x):
+    real = _clip_corr(x.real)
+    imag = _clip_corr(x.imag)
+    return real + 1j*imag
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: spaces around `*`

File: numba/tests/test_np_functions.py [hunk 4]
```diff
@@ -1265,74 +1277,133 @@ def test_cov_explicit_arguments(self):
...
+        self.cov_corrcoef_edge_cases(pyfunc, first_arg_name='x')
+
+        cfunc = jit(nopython=True)(pyfunc)
+        _check = partial(self._check_output, pyfunc, cfunc, abs_tol=1e-14)
+
+        for x in (np.nan, -np.inf, 3.142, 0):
+            params = {'x': x}
+            _check(params)
+
+    @unittest.skipUnless(np_version >= (1, 11), "corrcoef extreme value handling fixed Numpy 1.11+")
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: Long line.

File: numba/tests/test_np_functions.py [hunk 5]
```diff
@@ -1265,74 +1277,133 @@ def test_cov_explicit_arguments(self):
...
+    @unittest.skipUnless(np_version >= (1, 10), "corrcoef needs Numpy 1.10+")
     @needs_blas
-    def test_cov_edge_cases(self):
-        pyfunc = cov
+    def test_corrcoef_explicit_arguments(self):
+        pyfunc = corrcoef
+        cfunc = jit(nopython=True)(pyfunc)
+        _check = partial(self._check_output, pyfunc, cfunc, abs_tol=1e-14)
+
+        x = self.rnd.randn(1050).reshape(150, 7)
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: Does this array need to be this large?

File: docs/source/reference/numpysupported.rst (hunk 1 above)
Comment (rjenc29) [reply docs/source/reference/numpysupported.rst]: Added a few words.

File: numba/targets/arraymath.py (hunk 2 above)
Comment (rjenc29) [reply numba/targets/arraymath.py]: Done.

File: numba/targets/arraymath.py (hunk 3 above)
Comment (rjenc29) [reply numba/targets/arraymath.py]: Done.

File: numba/tests/test_np_functions.py (hunk 4 above)
Comment (rjenc29) [reply numba/tests/test_np_functions.py]: Shortened.

File: numba/tests/test_np_functions.py (hunk 5 above)
Comment (rjenc29) [reply numba/tests/test_np_functions.py]: Nope, made it 10x smaller.

**Comment by stuartarchibald** [comment]
Just restarted the Azure Pipeline CI as the failing error seemed utterly unrelated.

File: docs/source/reference/numpysupported.rst [hunk 6]
```diff
@@ -264,6 +264,7 @@ The following top-level functions are supported:
 * :func:`numpy.concatenate`
 * :func:`numpy.convolve` (only the 2 first arguments)
 * :func:`numpy.copy` (only the first argument)
+* :func:`numpy.corrcoef` (only the 3 first arguments, requires NumPy >= 1.10 and SciPy >= 0.16; extreme value handling per NumPy 1.11+)
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst]: Tiny thing... Any chance you could please put a line break in as this is a really long line now. I think the syntax is to line up the broken line with the first `:` of `:func` to get it to be a continued list item. Thanks.

**Comment by stuartarchibald** [comment]
Thanks for the fix-ups. There's a tiny edit to docs outstanding (long-line), ready to merge otherwise!

File: docs/source/reference/numpysupported.rst (hunk 6 above)
Comment (rjenc29) [reply docs/source/reference/numpysupported.rst]: Done - hopefully correctly (I have never had much luck building the docs).  Let me know if it's OK, cheers!

Comment (stuartarchibald) [reply docs/source/reference/numpysupported.rst]: Fix looks fine on inspection. Thanks.

**Pull request merged**